k8sfv/pod.go

```diff
@@ -0,0 +1,230 @@
+// Copyright (c) 2017 Tigera, Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"fmt"
+	"math/rand"
+	"strings"
+	"sync"
+
+	log "github.com/Sirupsen/logrus"
+	"github.com/containernetworking/cni/pkg/ns"
+	"github.com/projectcalico/libcalico-go/lib/backend/k8s"
+	"github.com/vishvananda/netlink"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/pkg/api/v1"
+)
+
+type podSpec struct {
+	// Actually there's no way to specify a pod's MAC address through the Kubernetes datastore.
+	// It's only used for an optimization anyway (to save an ARP request when routing on the
+	// last hop to a pod), so probably acceptable that we don't have it.
+	mac string
+	// IPv4 address.  If not provided, createPod will generate one.
+	ipv4Addr string
+	// IPv6 address.  Not yet supported by Kubernetes.
+	ipv6Addr string
+	// Pod name.  If not provided, createPod will generate one.
+	name string
+	// Labels (may be nil).
+	labels map[string]string
+}
+
+type localNetworking struct {
+	podIf     netlink.Link
+	hostIf    netlink.Link
+	namespace ns.NetNS
+}
+
+var localNetworkingMap = map[string]*localNetworking{}
+var localNetworkingMutex = sync.Mutex{}
+
+func createPod(clientset *kubernetes.Clientset, d deployment, nsName string, spec podSpec) {
+	name := spec.name
+	if name == "" {
+		name = fmt.Sprintf("run%d", rand.Uint32())
+	}
+	host := d.ChooseHost(clientset)
+	ip := spec.ipv4Addr
+	if ip == "" {
+		ip = GetNextAddr()
+	}
+	pod_in := &v1.Pod{
+		ObjectMeta: v1.ObjectMeta{Name: name},
+		Spec: v1.PodSpec{Containers: []v1.Container{{
+			Name:  fmt.Sprintf("container-%s", name),
+			Image: "ignore",
+		}},
+			NodeName: host.name,
+		},
+		Status: v1.PodStatus{
+			Phase: v1.PodRunning,
+			Conditions: []v1.PodCondition{{
+				Type:   v1.PodScheduled,
+				Status: v1.ConditionTrue,
+			}},
+			PodIP: ip,
+		},
+	}
+	if spec.labels != nil {
+		pod_in.ObjectMeta.Labels = spec.labels
+	}
+	log.WithField("pod_in", pod_in).Debug("Pod defined")
+	pod_out, err := clientset.Pods(nsName).Create(pod_in)
+	if err != nil {
+		panic(err)
+	}
+	log.WithField("pod_out", pod_out).Debug("Created pod")
+	pod_in = pod_out
+	pod_in.Status.PodIP = ip
+	pod_out, err = clientset.Pods(nsName).UpdateStatus(pod_in)
```

Comment:
For an accurate simulation, are we doing the same number of status updates as happen with a real pod?  Worth checking with @caseydavenport